app/Support/Web3NftHandler.php

```diff
@@ -39,149 +40,134 @@ public function __construct(
     /**
      * @param  Collection<int, Web3NftData>  $nfts
      */
-    public function store(Collection $nfts, bool $dispatchJobs = false): void
+    public function store(Collection $nfts): void
     {
-        $now = Carbon::now();
-
-        $nftsInCollection = collect($nfts)->groupBy('tokenAddress');
+        if ($nfts->isEmpty()) {
+            return;
+        }
 
-        $nftsGroupedByCollectionAddress = $nftsInCollection->map->first();
+        $now = Carbon::now();
 
-        $collectionsData = $nftsGroupedByCollectionAddress->flatMap(function (Web3NftData $nftData) use ($now, $nftsInCollection) {
-            $token = $nftData->token();
+        $collectionsData = $this->getCollections($nfts);
 
-            return [
-                $nftData->tokenAddress,
-                $nftData->networkId,
-                trim($nftData->collectionName),
-                $this->getSlug($nftData->collectionName),
-                $nftData->collectionSymbol,
-                $nftData->collectionDescription !== null ? strip_tags($nftData->collectionDescription) : null,
-                $nftData->collectionSupply,
-                $token ? $nftData->floorPrice?->price : null,
-                $token?->id,
-                $token ? $nftData->floorPrice?->retrievedAt : null,
-                json_encode([
-                    'image' => $nftData->collectionImage,
-                    'website' => $nftData->collectionWebsite,
-                    'socials' => $nftData->collectionSocials,
-                    'banner' => $nftData->collectionBannerImageUrl,
-                    'banner_updated_at' => $nftData->collectionBannerImageUrl ? $now : null,
-
-                ]),
-                $nftData->mintedBlock,
-                $nftData->mintedAt?->toDateTimeString(),
-                $this->lastRetrievedTokenNumber($nftsInCollection->get($nftData->tokenAddress)),
-                $now,
-                $now,
-            ];
-        });
+        /** @var string[] $columns */
+        $columns = array_keys($collectionsData->first());
 
-        $valuesPlaceholders = $nftsGroupedByCollectionAddress->map(fn () => '(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)')->join(',');
+        unset($columns['network_id']);
+        unset($columns['address']);
 
-        $ids = DB::transaction(function () use ($nfts, $collectionsData, $valuesPlaceholders, $dispatchJobs, $now) {
-            // upsert nfts/collections (if any)
-            if ($nfts->isEmpty()) {
-                return collect();
-            }
+        // We take all the columns (except the ones we unique on which are network_id and address)
+        // and generate `column = coalesce(excluded.column, collections.column)` for UPSERT...
+        $update = collect($columns)->mapWithKeys(fn ($column) => [
+            $column => DB::raw(sprintf('coalesce(excluded.%s, collections.%s)', $column, $column)),
```

Comment:
A few differences compared to previous version:

- we no longer trim the name, not sure if that's needed though

```
name = trim(coalesce(excluded.name, collections.name)),
```

- now we always keep a non-null floor price, while previously we'd potentially null it:

```
floor_price_token_id = excluded.floor_price_token_id,
floor_price_retrieved_at = excluded.floor_price_retrieved_at,
```

- this situation sounds similar to https://github.com/ArdentHQ/dashbrd/pull/98#discussion_r1328635128

so maybe we just never update it if it's already set ? 

i.e. something like 
`floor_price_token_id = coalesce(collections.floor_price_token_id, excluded.floor_price_token_id)`

Reply:
Regarding trimming, meh, we can just trim on the PHP level.

For the floor price being null, I'm kinda curious - why did we even null the price before? If the price is already set in the response, should we just set it in the database?

Reply:
> For the floor price being null, I'm kinda curious - why did we even null the price before? If the price is already set in the response, should we just set it in the database?

Likely just an oversight. It might even have been there before we added the floor price job for Mnemonic. But we should change it now while at it for consistency.

Reply:
> Regarding trimming, meh, we can just trim on the PHP level.

:+1:

Reply:
Hmm, with my PR, we already do take the first non-null value. Though, I suspect the order matters, right?

If both are set, will take from the database table:
```
floor_price_token_id = coalesce(collections.floor_price_token_id, excluded.floor_price_token_id)
```

If both are set, will take from the array to update:
```
floor_price_token_id = coalesce(excluded.floor_price_token_id, collections.floor_price_token_id)
```

Reply:
>  Though, I suspect the order matters, right?

Yes, we should use the former variant for this column. It will ensure we only write it if we don't have one yet, then it's consistent with https://github.com/ArdentHQ/dashbrd/pull/98#discussion_r1328635128 and only the actual floor price job is responsible to update the price afterwards.

Having an initial price means we have more likely something to show in the UI before the floor price job ran. That's the only reason why we should bother setting a price here.